### PR TITLE
428 - Added IdsAttributesMixin (to autowire attribute getters+setters)

### PR DESCRIPTION
**Pull request by clayinfor**
# Ids Attributes Mixin

This mixin adds attribute getter/setter helper functionality to components

**Related github/jira issue (required)**:

Closes https://github.com/infor-design/enterprise-wc/issues/428

**Steps necessary to review your pull request (required)**:

**Included in this Pull Request**:
- [ ] An e2e or functional test for the bug or feature.
- [ ] A note to the change log.

### CODE REVIEW

File: src/mixins/ids-attributes-mixin/ids-attributes-mixin.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+const TypeCasters = {
```
Comment (clayinfor) [inline src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:1]: This could be pluggable...

**Comment by tmcconechy** [comment]
I think would need to see how this looks in one of the converted components... 

Also debating if it should go in IdsElement or be a mixin (i guess if not all elements have attributes then a mixin does sense).

Didnt do a full review yet...

File: src/mixins/ids-attributes-mixin/ids-attributes-mixin.js (hunk 1 above)
Comment (clayinfor) [reply src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:1]:
```javascript
// IdsDataGrid.js could do something like this to override typecasters
get propTypeCasters() {
  return {
    ...super.propTypeCasters(),
    boolean: stringToBool,
    object: this.mySophisticatedDataParser,
  };
}
```

Comment (tmcconechy) [reply src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:1]: What would this be useful for? Not picturing it.

File: src/mixins/ids-attributes-mixin/ids-attributes-mixin.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  const propTypeCaster = target.propTypeCasters[typeof (defaultValue)] || Boolean;
+  const stringCaster = (typeof (defaultValue) === 'object') ? JSON.stringify : String;
+
+  target.setAttribute(attrName, stringCaster(defaultValue)); // initialize attribute
+
+  const getter = () => (propTypeCaster(target.getAttribute(attrName)) ?? defaultValue);
+  const setter = (value) => {
+    const stringValue = stringCaster(value);
+    changing?.bind(target)?.(stringValue);
+    target.setAttribute(attrName, stringValue);
```
Comment (EdwardCoyle) [inline src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:24]: @clayinfor I wonder if this handles boolean attributes?  Typically for those, you would want to fully remove the attribute if it were to become `false`, per the spec: https://github.com/infor-design/enterprise-wc/blob/main/doc/COMPONENTS.md#boolean-attributes

We might also want to talk about enum-style attributes, where they could either be `null` (no attribute), or be defined with a string type.  This is currently how the `colorVariant` property passed from IdsColorVariantMixin behaves: https://github.com/infor-design/enterprise-wc/blob/main/src/mixins/ids-color-variant-mixin/ids-color-variant-mixin.js#L63-L67

**Review by EdwardCoyle — COMMENTED** [review]
Just adding a comment on something we've run into regarding attributes in the past

File: src/mixins/ids-attributes-mixin/ids-attributes-mixin.js (hunk 1 above)
Comment (clayinfor) [reply src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:1]: It gives the component a way to override the default type-casters which as follows:
```javascript
const TypeCasters = {
  boolean: Boolean,
  number: Number,
  string: String,
  object: JSON.parse,
};
```

Comment (tmcconechy) [reply src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:1]: ok, not sure how to work it in but for object/array we might want to use this deepCopy instead https://github.com/infor-design/enterprise-wc/blob/main/src/components/ids-data-grid/ids-data-grid.js#L5 of JSON.parse as that could loose something...

Just putting it here for now

File: src/mixins/ids-attributes-mixin/ids-attributes-mixin.js (hunk 2 above)
Comment (clayinfor) [reply src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:24]: Thanks for pointing this out, @EdwardCoyle!  I think we should be able to easily implement that spec here.  

If, however, it turns out to be more complicated that I'm thinking, we could consider continuing to use `get/set exampleProp()` as we're currently doing.

File: src/mixins/ids-attributes-mixin/ids-attributes-mixin.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  const stringCaster = (typeof (defaultValue) === 'object') ? JSON.stringify : String;
+
+  target.setAttribute(attrName, stringCaster(defaultValue)); // initialize attribute
+
+  const getter = () => (propTypeCaster(target.getAttribute(attrName)) ?? defaultValue);
+  const setter = (value) => {
+    const stringValue = stringCaster(value);
+    before?.bind(target)?.(stringValue);
+    target.setAttribute(attrName, stringValue);
+    after?.bind(target)?.(stringValue);
```
Comment (jmacaluso711) [inline src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:25]: Just making a note here about the names of the callbacks.. Initially what came to mind for me was `beforeChange` and `changed` Maybe want to put it up for debate in a Teams survey or further discussion in next weeks wc meeting?

**Review by jmacaluso711 — COMMENTED** [review]
@clayinfor This is looking really great.. Left a comment about some of the callback naming conventions. Also wondering how much, if any, this could be improved/streamlined by adding typescript to our stack. I guess we'd need to do a cost-benefit analysis there.

File: src/mixins/ids-attributes-mixin/ids-attributes-mixin.js (hunk 3 above)
Comment (tmcconechy) [reply src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:25]: Id rather not go add typescript to the whole stack. That seems over kill to me.

**Comment by danielortiz1982** [comment]
@tmcconechy My outlook and opinion on implementation, I think this should be baked into `IdsElement` as it is core functionally that will built around the `attributeChangedCallback` and `observedAttributes` lifecycle functions. Just my thoughts on it. I feel like a mixin could work also.

File: src/mixins/ids-attributes-mixin/ids-attributes-mixin.js (hunk 3 above)
Comment (danielortiz1982) [reply src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:25]: @jmacaluso711 I like the idea, let Polly decided 😃

Comment (tmcconechy) [reply src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:25]: For this i like beforeChange and changed as well.

File: src/mixins/ids-attributes-mixin/ids-attributes-mixin.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+   * +      super();
+   * +
+   * +      this.prop(attributes.DATA, { fname: "John", lname: "Doe" });
+   * +      this.prop(attributes.AUTO_FOCUS);
+   * +
+   * +      // Reacting to prop changes using method-chaining
+   * +      this.prop(attributes.ALIGN_Y, 'top').before(this.alignYBefore).after(this.alignYAfter);
+   * +
+   * +      // Reacting to prop changes using the options argument
+   * +      this.prop(attributes.ALIGN_X, 'start', { after: this.alignXAfter });
```
Comment (tmcconechy) [inline src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:93]: I like this approach the best so far. Might want to see if we can group them so its less verbose in an array or object or method properties.

And if we do call it properties i think the mixin would become propertyMixin (we do have a concept of adding attributes like id's / data-automation-id) so might want to reserve that name

Comment (tmcconechy) [reply src/mixins/ids-attributes-mixin/ids-attributes-mixin.js:93]: But probably lets first make a decision on how the API looks first

**Comment by tmcconechy** [comment]
i agree @danielortiz1982 i did think at first it should be baked into IdsElement but the only thing is Not Every component does have attributes but if its that basic you probably just inherit from HTMLElement. So I guess i agree about baking it in as thats what i was originally thinking about this

**Comment by tmcconechy** [comment]
@clayinfor 

Discussed a bit with @lucacolumbu and we think if we can do it like this... it would be 💯 

```
export default class IdsBadge extends Base {
  @prop('disabled', Boolean)
  @prop('color', String)
  @prop('shape', Object, onShapeChanged, ...otheroptions)

  constructor() { // Not sure if has to be first
      super();
  }

  // Later
  onShapeChanged() {
  }
  
  // Fires after every prop is set
  onAllPropertiesSet() {
  }
```

I think the decorators can be there like this just need to workout the syntax... it it could go here https://github.com/infor-design/enterprise-wc/blob/main/src/core/ids-decorators.js

Feature Use Cases
- Only properties (no attributes)
- Fire an event only when the last property is set on loading
- Two settings depend on each other and fire changed when both are set

**Comment by EdwardCoyle** [comment]
Regarding this one:

> Only properties (no attributes)

Came to my mind that it could be handled this way... arguable whether or not "observed by attribute" is a default:
```js
@prop('shape', Object, onShapeChanged, {
  observed: false
})
```

**Comment by tmcconechy** [comment]
Just to confirm: i like Eds addition so put it in

**Comment by tmcconechy** [comment]
Will close this for now and can reraise

**Pull request closed without merging**